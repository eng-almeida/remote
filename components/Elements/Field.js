import React, { useState, useCallback } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { LabelOne } from '@/components/Elements/Typography'

const Label =  styled.label`
	${LabelOne}
	display: block;
	color: ${({ theme, isFocus }) => isFocus ? theme.colors.purple[200] : theme.colors.grays[100]};
	margin-bottom: ${({ theme }) => theme.space[1]};
	transition: color 0.2s;
`;

 const HelperText = styled.div`
	${LabelOne}
	color: ${({ theme }) => theme.colors.grays[300]};
	opacity: 0.5;
	margin-top: ${({ theme }) => theme.space[2]};
`;

const Field = ({ id, label, helperText, children }) => {
  const [isFocus, setFocus] = useState(false);
  const handleFocus = useCallback(() => setFocus(true), []);
  const handleBlur = useCallback(() => setFocus(false), []);

  return (
    <div>
      <Label isFocus={isFocus} htmlFor={id}>{label}</Label>
			{React.Children.map(children, child => (
				React.cloneElement(child, { onFocus: handleFocus, onBlur: handleBlur })
			))}
      <HelperText>{helperText}</HelperText>
    </div>
  );
}

Field.propTypes = {
	id: PropTypes.string.isRequired,
	label: PropTypes.string.isRequired,
	helperText: PropTypes.string.isRequired,
	children: PropTypes.element.isRequired,
};

export default Field;