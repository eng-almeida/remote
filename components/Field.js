import React, { useState, useCallback } from 'react';
import styled from 'styled-components';

const Label =  styled.label`
	display: block;
	color: ${({ theme, isFocus }) => isFocus ? theme.colors.purple[200] : theme.colors.grays[100]};
	font-size: 13px;
	line-height: 16px;
	margin-bottom: ${({ theme }) => theme.space[1]};
`;

 const HelperText = styled.div`
	color: ${({ theme }) => theme.colors.grays[300]};
	opacity: 0.5;
	font-size: 13px;
	line-height: 16px;
	margin-top: ${({ theme }) => theme.space[2]};
`;

const Field = ({ id, label, helperText, children }) => {
  const [isFocus, setFocus] = useState(false);
  const handleFocus = useCallback(() => setFocus(true), []);
  const handleBlur = useCallback(() => setFocus(false), []);

  return (
    <div>
      <Label isFocus={isFocus} htmlFor={id}>{label}</Label>
				{React.Children.map(children, child => (
					React.cloneElement(child, { onFocus: handleFocus, onBlur: handleBlur })
				))}
      <HelperText>{helperText}</HelperText>
    </div>
  )
}

export default Field;