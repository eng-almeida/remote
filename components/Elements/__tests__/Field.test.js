import { screen, fireEvent } from '@testing-library/react';
import Field from '@/components/Elements/Field';

describe('Field component', () => {
	it('should match snapshot', () => {
		const { container } = renderWithProviders(<Field id="test" label="Test"><input id="test" /></Field>);
		expect(container).toMatchSnapshot();
	});

	describe('when input is focused', () => {
		it('should match snapshot', () => {
			const { container } = renderWithProviders(<Field id="test" label="Test"><input id="test" /></Field>);
			const input = screen.getByLabelText(/Test/i);
			fireEvent.focus(input);
			expect(container).toMatchSnapshot();
		});
	});
})