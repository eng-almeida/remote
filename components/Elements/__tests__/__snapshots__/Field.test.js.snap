// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Field component should match snapshot 1`] = `
.c0 {
  font-size: 13px;
  line-height: 16px;
  display: block;
  color: #778CA3;
  margin-bottom: 4px;
}

.c1 {
  font-size: 13px;
  line-height: 16px;
  color: #00234B;
  opacity: 0.5;
  margin-top: 8px;
}

<div>
  <div>
    <label
      class="c0"
      for="test"
    >
      Test
    </label>
    <input
      id="test"
    />
    <div
      class="c1"
    />
  </div>
</div>
`;

exports[`Field component when input is focused should match snapshot 1`] = `
.c0 {
  font-size: 13px;
  line-height: 16px;
  display: block;
  color: #624DE3;
  margin-bottom: 4px;
}

.c1 {
  font-size: 13px;
  line-height: 16px;
  color: #00234B;
  opacity: 0.5;
  margin-top: 8px;
}

<div>
  <div>
    <label
      class="c0"
      for="test"
    >
      Test
    </label>
    <input
      id="test"
    />
    <div
      class="c1"
    />
  </div>
</div>
`;
